django/contrib/admin/templates/admin/includes/fieldset.html

```diff
@@ -1,5 +1,11 @@
-<fieldset class="module aligned {{ fieldset.classes }}">
-    {% if fieldset.name %}<h2>{{ fieldset.name }}</h2>{% endif %}
+{% with prefix=fieldset.formset.prefix|default:"fieldset" id_prefix=id_prefix|default:"0" id_suffix=id_suffix|default:"0" name=fieldset.name|default:""|slugify %}
+<fieldset class="module aligned {{ fieldset.classes }}" {% if name %}aria-labelledby="{{ prefix }}-{{ id_prefix}}-{{ name }}-{{ id_suffix }}"{% endif %}>
+    {% if name %}
+        {% if fieldset.is_collapsible %}<details><summary>{% endif %}
+        <h{{ heading_level|default:2 }} id="{{ prefix }}-{{ id_prefix}}-{{ name }}-{{ id_suffix }}" class="fieldset-heading">{{ fieldset.name }}</h{{ heading_level|default:2 }}>
```

Comment:
I think either this id is too complex or the previously formed id `{{ inline_admin_formset.formset.prefix }}-heading` is too simple. Maybe `{{ fieldset.formset.prefix }}-fieldset-header-{{ id_suffix }}` would be more consistent.

This logic is also missing tests such as
- checking for id uniqueness i.e. `self.assertContains(response, `id="this-id-thing"`, html=True, count=1)`
- checking for `aria-labelledby`

Reply:
Thank you! I agree that tests are missing, I will add some, which will definitely help validate what I expose down below. What follows may be confusing though I tried to be as clear as possible, but the summary is that I'll build a test case with the scenario below and proper asserts.

> I think either this id is too complex or the previously formed id `{{ inline_admin_formset.formset.prefix }}-heading` is too simple.

Regarding the complexity of the proposed `id`, and given that the `id` attr needs to have a unique value, we need to consider the following scenarios where the `fieldset.html` is included in other templates:
1. via "toplevel" inclusion in the `change_form.html` (https://github.com/django/django/blob/main/django/contrib/admin/templates/admin/change_form.html#L48)
2. via a nested for loop via `StackedInline` in `stacked.html` (https://github.com/django/django/blob/main/django/contrib/admin/templates/admin/edit_inline/stacked.html#L22)

>  Maybe `{{ fieldset.formset.prefix }}-fieldset-header-{{ id_suffix }}` would be more consistent.

Sadly, `{{ fieldset.formset.prefix }}-fieldset-header-{{ id_suffix }}` is not truly unique, and as far as my research goes, the "only" way to built a truly unique `id` for the fieldset heading is using all of the following elements (in all cases, the fieldset name is accessory in that it does not provide uniqueness but it helps building a nicer `id`):

1. For item 1, where there is no formset associated with the fieldset, the fieldset name and the for loop counter suffices
2. For item 2, we need to consider:
    * the formset prefix (which is also used as part of most `id`s in `stacked.html`), since in an admin page there can be X instances of inlines, so that would mean there are X different formset prefixes
    * the for loop counter that iterates over each formset (since there can be Y `inline_admin_form` inside each formset), this is what the proposal defines as `parent_counter` in `stacked.html` and is used as `id_prefix` in `fieldset.html`)
    * the for loop counter that iterates over each fieldset in each form in each formset (since there can be Z fieldsets in each inline_admin_form), this is what the proposal uses as `id_suffix` in `fieldset.html`
    * As with item 1, the fieldset name is used to make the `id` a little nicer, not because it adds uniqueness

So for models and admins like the ones defined at the bottom of this comment, we get fieldset headers like this (the ids are of the form `cover_set-X-Y-details-Z`, where `coverset-X` is the formset prefix, Y and X are as described above, and `details` is the fieldset name):
```html
        <!-- these two are toplevel fieldsets, defined at the admin model level -->
        <h2 id="fieldset-0-advanced-options-1" class="fieldset-heading">Advanced options</h2>
        <h2 id="fieldset-0-advanced-options-2" class="fieldset-heading">Advanced options</h2>

  <h2 id="cover_set-heading" class="inline-heading"> <!-- this one `CoverTabularInline` instance -->
  <h2 id="cover_set-2-heading" class="inline-heading"> <!-- this the second `CoverTabularInline` instance -->

  <h2 id="cover_set-3-heading" class="inline-heading"> <!-- this is the `CoverStackedExtra2Inline` instance -->
        <h4 id="cover_set-3-0-details-1" class="fieldset-heading">Details</h4> <!-- first fielset of the first form -->
        <h4 id="cover_set-3-0-details-2" class="fieldset-heading">Details</h4> <!-- second fielset of the first form -->
        <h4 id="cover_set-3-1-details-1" class="fieldset-heading">Details</h4> <!-- first fielset of the second form -->
        <h4 id="cover_set-3-1-details-2" class="fieldset-heading">Details</h4> <!-- etc. -->
        <h4 id="cover_set-3-2-details-1" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-3-2-details-2" class="fieldset-heading">Details</h4>

  <h2 id="cover_set-4-heading" class="inline-heading"> <!-- this is the `CoverStackedExtra3Inline` instance -->
        <h4 id="cover_set-4-0-details-1" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-0-details-2" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-1-details-1" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-1-details-2" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-2-details-1" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-2-details-2" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-3-details-1" class="fieldset-heading">Details</h4>
        <h4 id="cover_set-4-3-details-2" class="fieldset-heading">Details</h4>
```
* models.py
```python
from django.db import models
from django.utils.timezone import now


class Book(models.Model):
    title = models.CharField(max_length=100)
    author = models.CharField(max_length=100)
    publisher = models.CharField(max_length=100)
    creation_date = models.DateField(default=now)
    update_date = models.DateField(default=now)
    publication_date = models.DateField(default=now)

    def __str__(self):
        return self.title


class Cover(models.Model):
    book = models.ForeignKey(Book, on_delete=models.CASCADE)
    image = models.CharField(max_length=100)
    title = models.CharField(max_length=100)
    description = models.TextField()
    creation_date = models.DateField(default=now)
    update_date = models.DateField(default=now)
    updated_by = models.CharField(max_length=100)
```
* admin.py
```python
from django.db import models
from django.contrib import admin

from .models import Book, Cover


class CoverInlineMixin:
    model = Cover
    extra = 2
    fieldsets = (
        (None, {"fields": ("image", "title")}),
        ("Details", {
            "fields": ("description", "creation_date"),
        }),
        ("Details", {  # Name intentionally duplicated
            "fields": ("update_date", "updated_by"),
        }),
    )


class CoverTabularInline(CoverInlineMixin, admin.TabularInline):
    pass


class CoverStackedExtra2Inline(CoverInlineMixin, admin.StackedInline):
    pass


class CoverStackedExtra3Inline(CoverInlineMixin, admin.StackedInline):
    extra = 3


class BookAdmin(admin.ModelAdmin):
    list_display = ("title", "author", "publisher","publication_date")
    search_fields = ("title", "author")
    list_filter = ("author", "publication_date")
    fieldsets = (
        (None, {
            "fields": ("title", "author")
        }),
        ("Advanced options", {
            "fields": ("publisher", "publication_date",)
        }),
        ("Advanced options", {
            "fields": ("creation_date", "update_date",)
        }),
    )
    inlines = [
        CoverTabularInline,
        CoverTabularInline,
        CoverStackedExtra2Inline,
        CoverStackedExtra3Inline,
    ]


admin.site.register(Book, BookAdmin)
```

Reply:
@sarahboyce I have pushed a complete test for the headings, including "h-level" and id. I realized the heading needed a "-heading" suffix to be consistent with other ids, but otherwise the ID is as described above. Resolving but let me know if you would like something changed! Thank you.